modules/sockets/sockets.cpp

```diff
@@ -15,233 +11,370 @@
-    // Now we recieve
-    tmp = recv(socket, tmpchar, length-1, 0);
-	if (tmp == -1)
+	int sockfd = params[1];
+	int length = params[3];
+
+	char *recv_buffer = new char[length];
+
+	if(recv_buffer == nullptr)
+		return -1;
+
+	memset(recv_buffer, 0, length);
+
+	int bytes_received = -1;
+	bytes_received = recv(sockfd, recv_buffer, length - 1, 0);
+
+	if(bytes_received == -1)
 	{
-		delete [] tmpchar;
+		delete[] recv_buffer;
 		return -1;
 	}
-    // And put a copy of our recieved data into amx's string
-    tmpchar[tmp]='\0';
-    int nlen = 0;
-    //int max = params[3];
-    int max = length-1;
-    const char* src = tmpchar;
-    cell* dest = MF_GetAmxAddr(amx,params[2]);
-    while(max--&&nlen<tmp){
-     *dest++ = (cell)*src++;
-     nlen++;
-    }
-    *dest = 0;
-    // And we need to free up the space to avoid wasting memory
-    delete [] tmpchar;
-    // And finnally, return the what recv returnd
-    return tmp;
+
+	recv_buffer[bytes_received] = '\0';
+
+	cell* destination = MF_GetAmxAddr(amx, params[2]);
+	strncopy(destination, recv_buffer, length - 1);
+
+	delete[] recv_buffer;
+
+	return bytes_received;
 }
 
 // native socket_send(_socket, _data[], _length);
-static cell AMX_NATIVE_CALL socket_send(AMX *amx, cell *params)  /* 3 param */
+static cell AMX_NATIVE_CALL socket_send(AMX *amx, cell *params)
 {
-	// We get the string from amx
-	int len;
-	int socket = params[1];
-	char* data = MF_GetAmxString(amx,params[2],0,&len);
-	// And send it to the socket
-	return send(socket, data, len, 0);
-}
+	int sockfd = params[1];
+	int length = 0;
+	
+	char *data = MF_GetAmxString(amx, params[2], 0, &length);
 
-static char *g_buffer = NULL;
-static size_t g_buflen = 0;
+	return send(sockfd, data, length, 0);
+}
 
 // native socket_send2(_socket, _data[], _length);
-static cell AMX_NATIVE_CALL socket_send2(AMX *amx, cell *params)  /* 3 param */
+static cell AMX_NATIVE_CALL socket_send2(AMX *amx, cell *params)
 {
-	// We get the string from amx
-	int len = params[3];
-	int socket = params[1];
-	if ((size_t)len > g_buflen)
+	int sockfd = params[1];
+	int length = params[3];
+	
+	if(length > g_send2_buffer_length)
 	{
-		delete [] g_buffer;
-		g_buffer = new char[len+1];
-		g_buflen = len;
+		delete[] g_send2_buffer;
+
+		g_send2_buffer = new char[length + 1];
+		g_send2_buffer_length = length;
 	}
 
-	cell *pData = MF_GetAmxAddr(amx, params[2]);
-	char *pBuffer = g_buffer;
+	cell *data = MF_GetAmxAddr(amx, params[2]);
+
+	while(length--)
+		*g_send2_buffer++ = (char)*data++;
+
+	return send(sockfd, g_send2_buffer, length, 0);
+}
+
+// native socket_change(_socket, _timeout = 100000);
+static cell AMX_NATIVE_CALL socket_change(AMX *amx, cell *params)
+{
+	int sockfd = params[1];
+	unsigned int timeout = params[2];
+
+	struct timeval tv;
+	tv.tv_sec = 0;
+	tv.tv_usec = timeout;
+
+	fd_set readfds;
+	FD_ZERO(&readfds);
+	FD_SET(sockfd, &readfds);
+
+	return (select(sockfd + 1, &readfds, nullptr, nullptr, &tv) > 0) ? 1 : -1;
```

Comment:
```/**
 * Receives data.
 *
 * @note The ammount of bytes than you end up receiving can be less than the one you expected
 *
 * @param _socket    Socket descriptor
 * @param _data      Array to save the data
 * @param _length    Length of the array
 *
 * @return           Ammount of bytes received
 *                   0 if the peer closed the connection
 *                   -1 on failure
 */
native socket_recv(_socket, _data[], _length);
```

Every return value but this one can be changed to 0 without any problem.

Changing socket_recv() to 0 would remove the possibility of knowing if the peer closed the connection.

Should we remove this feature in order to keep everything consistent ? I'm not sure it would be of any use, and the socket's status can be check using other functions.